src/hats_import/verification/run_verification.py

```diff
@@ -1,14 +1,439 @@
 
-    # implement everything else.
-    raise NotImplementedError("Verification not yet implemented.")
+    verifier = Verifier.from_args(args)
+    verifier.run(write_mode=write_mode, check_metadata=check_metadata)
+
+    return verifier
+
+
+def now() -> str:
+    """Get the current time as a string."""
+    return datetime.datetime.now(datetime.timezone.utc).strftime("%Y/%m/%d %H:%M:%S %Z")
+
+
+@dataclass(kw_only=True, frozen=True)
+class Result:
+    """Verification test result for a single test."""
+
+    datetime: str = field(default_factory=now)
+    """The date and time when the test was run."""
+    passed: bool = field()
+    """Whether the test passed."""
+    test: str = field()
+    """Test name."""
+    target: str = field()
+    """The file(s) targeted by the test."""
+    description: str = field()
+    """Test description."""
+    bad_files: list[str] = field(default_factory=list)
+    """List of additional files that caused the test to fail (empty if none or not applicable)."""
+
+
+@dataclass(kw_only=True)
+class Verifier:
+    """Run verification tests. To create an instance of this class, use `Verifier.from_args`."""
+
+    args: VerificationArguments = field()
+    """Arguments to use during verification."""
+    metadata_ds: pyarrow.dataset.Dataset = field()
+    """Pyarrow dataset, loaded from the _metadata file."""
+    files_ds: pyarrow.dataset.Dataset = field()
+    """Pyarrow dataset, loaded from the parquet data files."""
+    common_metadata_schema: pyarrow.Schema = field()
+    """Pyarrow schema, loaded from the _common_metadata file."""
+    constructed_truth_schema: pyarrow.Schema = field()
+    """Pyarrow schema treated as truth during verification. This is constructed
+    from `common_metadata_schema` and `args.truth_schema`. `common_metadata_schema`
+    is used for hats-specific columns. If provided, `args.truth_schema` is used
+    for all other columns plus metadata, otherwise `common_metadata_schema` is used.
+    """
+    results: list[Result] = field(default_factory=list)
+    """List of results, one for each test that has been done."""
+
+    @classmethod
+    def from_args(cls, args: VerificationArguments) -> "Verifier":
+        """Create a `Verifier` with initialized datasets and schemas based on `args`.
+
+        Parameters
+        ----------
+        args : VerificationArguments
+            Arguments for the Verifier.
+
+        Returns
+        -------
+        Verifier
+        """
+        args.output_path.mkdir(exist_ok=True, parents=True)
+
+        print("Loading dataset and schema.")
+        files_ds = pyarrow.dataset.dataset(args.input_dataset_path)
+        metadata_ds = pyarrow.dataset.parquet_dataset(
+            hats.io.paths.get_parquet_metadata_pointer(args.input_catalog_path)
+        )
+
+        input_truth_schema = None
+        if args.truth_schema is not None:
+            input_truth_schema = pyarrow.dataset.parquet_dataset(args.truth_schema).schema
```

Comment:
```suggestion
            input_truth_schema = pds.parquet_dataset(args.truth_schema).schema
```

As well as other places in this file it appears. We use the `pds` convention for pyarrow dataset, and would improve readability on these lines.